dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp

```diff
@@ -3755,7 +3787,10 @@ RtpsUdpDataLink::RtpsWriter::gather_heartbeats(OPENDDS_VECTOR(TransportQueueElem
   if (!preassociation_readers_.empty()) {
     for (ReaderInfoSet::const_iterator pos = preassociation_readers_.begin(), limit = preassociation_readers_.end();
          pos != limit; ++pos) {
-      gather_preassociation_heartbeat_i(meta_submessages, meta_submessage, *pos);
+      const ReaderInfo_rch& reader = *pos;
+      gather_preassociation_heartbeat_i(meta_submessages, meta_submessage, reader);
+      reader->expecting_ack_ = true;
+      ++link->expected_acks_;
```

Comment:
Lock link->heartbeat_mutex_? Or add a helper method for it? Same for increments below.